core/clustersmngr/client.go

```diff
@@ -0,0 +1,171 @@
+package clustersmngr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Client thin wrapper to controller-runtime/client  adding multi clusters context.
+type Client interface {
+	// Get retrieves an obj for the given object key.
+	Get(ctx context.Context, cluster string, key client.ObjectKey, obj client.Object) error
+	// List retrieves list of objects for a given namespace and list options.
+	List(ctx context.Context, cluster string, list client.ObjectList, opts ...client.ListOption) error
+
+	// Create saves the object obj.
+	Create(ctx context.Context, cluster string, obj client.Object, opts ...client.CreateOption) error
+	// Delete deletes the given obj
+	Delete(ctx context.Context, cluster string, obj client.Object, opts ...client.DeleteOption) error
+	// Update updates the given obj.
+	Update(ctx context.Context, cluster string, obj client.Object, opts ...client.UpdateOption) error
+	// Patch patches the given obj
+	Patch(ctx context.Context, cluster string, obj client.Object, patch client.Patch, opts ...client.PatchOption) error
+
+	// ClusteredList retrieves list of objects for all clusters.
+	ClusteredList(ctx context.Context, clist ClusteredObjectList, opts ...client.ListOption) error
+
+	// ClientsPool returns the clients pool.
+	ClientsPool() ClientsPool
+}
+
+type ClusterNotFoundError struct {
+	Cluster string
+}
+
+func (e ClusterNotFoundError) Error() string {
+	return fmt.Sprintf("cluster=%s not found", e.Cluster)
+}
+
+type clustersClient struct {
+	pool ClientsPool
+}
+
+func NewClient(clientsPool ClientsPool) Client {
+	return &clustersClient{
+		pool: clientsPool,
+	}
+}
+
+func (c *clustersClient) ClientsPool() ClientsPool {
+	return c.pool
+}
+
+func (c *clustersClient) Get(ctx context.Context, cluster string, key client.ObjectKey, obj client.Object) error {
+	client := c.pool.Clients()[cluster]
```

Comment:
that `c.pool.Clients()[cluster]` hurts my eyes a bit, totally personal taste. I know it does not fail, but my brain constantly says "it should fail if the key is not defined".

```go
func (cp *clientsPool) Client(name string) (client.Client, error) {
  if c, found := cp.clients[name]; found && c != nil {
    return c, nil
  }

  return ClusterNotFoundError{Cluster: name}
}

client, err := c.pool.Client(cluster)
if err != nil {
  return err
}
```

Not a blocker at all. That's only an extra sugar.

Reply:
agree, much better, pushing the change